File: OpenMP/OpenUH_OpenMP_Validation/FORTRAN/omp_task_untied.f

      INCLUDE "omp_my_sleep.f"

      INTEGER FUNCTION test_omp_task_untied()
        IMPLICIT NONE
        INCLUDE "omp_testsuite.f"
        
        EXTERNAL my_sleep
        INTEGER omp_get_num_threads, omp_get_thread_num
        INTEGER myj
        INTEGER i,j
        INTEGER cnt
        INTEGER start_tid(NUM_TASKS)
        INTEGER current_tid(NUM_TASKS)
        COMMON /orphvars/ j, cnt, start_tid, current_tid
        

        cnt = 0
        do i = 1, NUM_TASKS
          start_tid(i) = 0
          current_tid(i) = 0
        end do

!$omp parallel private(myj) shared(j)
!$omp single
        do i=1, NUM_TASKS
        j = i
        
        myj = j
!$omp task untied
          call my_sleep(SLEEPTIME)
          start_tid(myj) = omp_get_thread_num()
!$omp taskwait
      if (MOD(start_tid(myj),2) .ne. 0) then
        call my_sleep(SLEEPTIME)
        current_tid(myj) = omp_get_thread_num()
      
       else
        current_tid(myj) = omp_get_thread_num()
       end if
!$omp end task
        
        end do
!$omp end single
!$omp end parallel

        test_omp_task_untied = 0

        ! check if at least one untied task switched threads
        do i=1, NUM_TASKS
          if (current_tid(i) .ne. start_tid(i)) then
               test_omp_task_untied = 1
          end if
        end do

      END FUNCTION
!This is the main driver to invoke different test functions
      SUBROUTINE mpc_user_main
      IMPLICIT NONE
      INTEGER failed, success !Number of failed/succeeded tests
      INTEGER num_tests,crosschecked, crossfailed, j
      INTEGER temp,temp1
      INCLUDE "omp_testsuite.f"

      INTEGER test_omp_task_untied


      CHARACTER*50 logfilename !Pointer to logfile
      INTEGER result 

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
!      WRITE (*,*) "Enter logFilename:" 
!      READ  (*,*) logfilename

 
      WRITE (*,*) "######## OpenMP Validation Suite V 3.0a ######"
      WRITE (*,*) "## Repetitions:", N 
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      crossfailed=0
      result=1
      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Testing omp task untied"
      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) 
      WRITE (*,*) "testname: test_omp_task_untied"
      WRITE (*,*) "(Crosstests should fail)"
      WRITE (*,*)
      
      DO j = 1, N
        temp =  test_omp_task_untied()
        IF (temp .EQ. 1) THEN
          WRITE (*,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (*,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO

      
      IF (failed .EQ. 0) THEN
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END SUBROUTINE 

File: OpenMP/OpenUH_OpenMP_Validation/C/omp_task_untied.c

#include <stdio.h>
#include <math.h>
#include "omp_testsuite.h"
#include "omp_my_sleep.h"

int test_omp_task_untied(FILE * logFile){

  
  int i;
  int count;
  int start_tid[NUM_TASKS];
  int current_tid[NUM_TASKS];
  
  count = 0;
  
  /*initialization*/
  for (i=0; i< NUM_TASKS; i++){
    start_tid[i]=0;
    current_tid[i]=0;
  }
  
  #pragma omp parallel firstprivate(i)
  {
    #pragma omp single
    {
      for (i = 0; i < NUM_TASKS; i++) {
        
        int myi = i;
        #pragma omp task untied
        {
          my_sleep(SLEEPTIME);
          start_tid[myi] = omp_get_thread_num();
          current_tid[myi] = omp_get_thread_num();
          
          #pragma omp taskwait
          
          if((start_tid[myi] %2) !=0){
            my_sleep(SLEEPTIME);
            current_tid[myi] = omp_get_thread_num();
          
          } /* end of if */ 
          else {
            current_tid[myi] = omp_get_thread_num();
          }
          

        } /*end of omp task */
        
      } /* end of for */
    } /* end of single */
  } /* end of parallel */

  for (i=0;i<NUM_TASKS; i++)
  {
    printf("start_tid[%d]=%d, current_tid[%d]=%d\n",i, start_tid[i], i , current_tid[i]);
    if (current_tid[i] == start_tid[i])
      count++;
  }
  return (count<NUM_TASKS);
}
int main()
{
	int i;			/* Loop index */
	int result;		/* return value of the program */
	int failed=0; 		/* Number of failed tests */
	int success=0;		/* number of succeeded tests */


	printf("######## OpenMP Validation Suite V %s ######\n", OMPTS_VERSION );
	printf("## Repetitions: %3d                       ####\n",REPETITIONS);
	printf("## Loop Count : %6d                    ####\n",LOOPCOUNT);
	printf("##############################################\n");
	printf("Testing omp task untied\n\n");

	fprintf(stdout,"######## OpenMP Validation Suite V %s ######\n", OMPTS_VERSION );
	fprintf(stdout,"## Repetitions: %3d                       ####\n",REPETITIONS);
	fprintf(stdout,"## Loop Count : %6d                    ####\n",LOOPCOUNT);
	fprintf(stdout,"##############################################\n");
	fprintf(stdout,"Testing omp task untied\n\n");

	for ( i = 0; i < REPETITIONS; i++ ) {
		fprintf (stdout, "\n\n%d. run of test_omp_task_untied out of %d\n\n",i+1,REPETITIONS);
		if(test_omp_task_untied(stdout)){
			fprintf(stdout,"Test successful.\n");
			success++;
		}
		else {
			fprintf(stdout,"Error: Test failed.\n");
			printf("Error: Test failed.\n");
			failed++;
		}
	}

    if(failed==0){
		fprintf(stdout,"\nDirective worked without errors.\n");
		printf("Directive worked without errors.\n");
		result=0;
	}
	else{
		fprintf(stdout,"\nDirective failed the test %i times out of %i. %i were successful\n",failed,REPETITIONS,success);
		printf("Directive failed the test %i times out of %i.\n%i test(s) were successful\n",failed,REPETITIONS,success);
		result = (int) (((double) failed / (double) REPETITIONS ) * 100 );
	}
	printf ("Result: %i\n", result);
	return result;
}
